File: tests/atomic.test.js
import { describe, test, expect, beforeEach, afterEach } from 'bun:test';
import fs from 'fs';
import path from 'path';
import { atomicWriteSync, atomicWriteJsonSync } from '../cli/lib/atomic.js';

const TMP = path.join(import.meta.dir, '.tmp-atomic');

beforeEach(() => {
  fs.mkdirSync(TMP, { recursive: true });
});

afterEach(() => {
  fs.rmSync(TMP, { recursive: true, force: true });
});

describe('atomicWriteSync', () => {
  test('writes content to file', () => {
    const file = path.join(TMP, 'test.txt');
    atomicWriteSync(file, 'hello world');
    expect(fs.readFileSync(file, 'utf8')).toBe('hello world');
  });

  test('overwrites existing file', () => {
    const file = path.join(TMP, 'test.txt');
    atomicWriteSync(file, 'first');
    atomicWriteSync(file, 'second');
    expect(fs.readFileSync(file, 'utf8')).toBe('second');
  });

  test('creates parent directories', () => {
    const file = path.join(TMP, 'nested', 'dir', 'test.txt');
    atomicWriteSync(file, 'deep');
    expect(fs.readFileSync(file, 'utf8')).toBe('deep');
  });

  test('no temp files left behind', () => {
    const file = path.join(TMP, 'clean.txt');
    atomicWriteSync(file, 'data');
    const files = fs.readdirSync(TMP);
    expect(files).toEqual(['clean.txt']);
  });
});

describe('atomicWriteJsonSync', () => {
  test('writes pretty JSON with trailing newline', () => {
    const file = path.join(TMP, 'data.json');
    atomicWriteJsonSync(file, { key: 'value', nested: { a: 1 } });
    const content = fs.readFileSync(file, 'utf8');
    expect(content).toBe(JSON.stringify({ key: 'value', nested: { a: 1 } }, null, 2) + '\n');
  });

  test('writes valid JSON that can be parsed back', () => {
    const file = path.join(TMP, 'roundtrip.json');
    const data = { mintInvoked: true, task: 'test', mode: 'quick' };
    atomicWriteJsonSync(file, data);
    const parsed = JSON.parse(fs.readFileSync(file, 'utf8'));
    expect(parsed).toEqual(data);
  });
});

File: cli/internal/atomic/atomic_test.go
package atomic

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

func TestWrite(t *testing.T) {
	dir := t.TempDir()
	file := filepath.Join(dir, "test.txt")

	if err := WriteString(file, "hello world"); err != nil {
		t.Fatal(err)
	}
	assertFile(t, file, "hello world")

	if err := WriteString(file, "second"); err != nil {
		t.Fatal(err)
	}
	assertFile(t, file, "second")

	nested := filepath.Join(dir, "nested", "dir", "test.txt")
	if err := WriteString(nested, "deep"); err != nil {
		t.Fatal(err)
	}
	assertFile(t, nested, "deep")

	clean := filepath.Join(dir, "clean.txt")
	if err := WriteString(clean, "data"); err != nil {
		t.Fatal(err)
	}
	entries, err := os.ReadDir(dir)
	if err != nil {
		t.Fatal(err)
	}
	for _, entry := range entries {
		if entry.Type().IsRegular() && entry.Name() != "test.txt" && entry.Name() != "clean.txt" {
			t.Fatalf("unexpected temp file left behind: %s", entry.Name())
		}
	}
}

func TestWriteJSON(t *testing.T) {
	dir := t.TempDir()
	file := filepath.Join(dir, "data.json")
	data := map[string]any{"key": "value", "nested": map[string]any{"a": float64(1)}}
	if err := WriteJSON(file, data); err != nil {
		t.Fatal(err)
	}
	assertFile(t, file, "{\n  \"key\": \"value\",\n  \"nested\": {\n    \"a\": 1\n  }\n}\n")

	var parsed map[string]any
	b, err := os.ReadFile(file)
	if err != nil {
		t.Fatal(err)
	}
	if err := json.Unmarshal(b, &parsed); err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(parsed, data) {
		t.Fatalf("parsed JSON = %#v, want %#v", parsed, data)
	}
}

func assertFile(t *testing.T, path, want string) {
	t.Helper()
	b, err := os.ReadFile(path)
	if err != nil {
		t.Fatal(err)
	}
	if string(b) != want {
		t.Fatalf("%s = %q, want %q", path, string(b), want)
	}
}
